File: frontend/src/types/task.ts
import { Platform, CrawlerType, TaskStatus } from './common'
import { CrawlerConfig } from './crawler'

export interface Task {
  id: string
  name: string
  platform: Platform
  crawlerType: CrawlerType
  status: TaskStatus
  config: CrawlerConfig
  progress: TaskProgress
  createdAt: string
  startedAt?: string
  completedAt?: string
  errorMessage?: string
  logs?: TaskLog[]
}

export interface TaskProgress {
  current: number
  total: number
  percentage: number
  speed?: number // items per second
  eta?: number // estimated time remaining in seconds
}

export interface TaskLog {
  timestamp: string
  level: 'info' | 'warning' | 'error' | 'success'
  message: string
}

export interface TaskStatistics {
  totalTasks: number
  runningTasks: number
  completedTasks: number
  failedTasks: number
  pendingTasks: number
}

export interface CreateTaskRequest {
  name: string
  platform: Platform
  crawlerType: CrawlerType
  config: CrawlerConfig
}

export interface TaskActionRequest {
  action: 'start' | 'pause' | 'resume' | 'cancel'
}

// WebSocket event types
export interface WebSocketEvent {
  type: 'task_started' | 'task_progress' | 'task_completed' | 'task_error' | 'task_log'
  taskId: string
  data: any
  timestamp: string
}

export interface TaskProgressEvent extends WebSocketEvent {
  type: 'task_progress'
  data: TaskProgress
}

export interface TaskLogEvent extends WebSocketEvent {
  type: 'task_log'
  data: TaskLog
}

File: backend/app/schemas/task.py
# -*- coding: utf-8 -*-
# Copyright (c) 2025 relakkes@gmail.com
#
# This file is part of MediaCrawler project.

"""Task-related schemas"""

from datetime import datetime
from typing import Optional, Any
from pydantic import BaseModel, Field


class TaskProgress(BaseModel):
    """Task progress information"""
    task_id: str = Field(alias="taskId")
    progress: int = Field(description="Progress percentage 0-100")
    status: str = Field(description="Current task status")
    items_collected: int = Field(alias="itemsCollected")
    estimated_remaining: Optional[int] = Field(alias="estimatedRemaining", default=None)
    speed: Optional[float] = Field(default=None, description="Items per minute")
    current_page: Optional[int] = Field(alias="currentPage", default=None)
    
    class Config:
        populate_by_name = True


class TaskLog(BaseModel):
    """Task log entry"""
    timestamp: str
    level: str = Field(description="INFO|WARNING|ERROR|SUCCESS")
    message: str


class TaskBase(BaseModel):
    """Base task schema"""
    platform: str = Field(description="Platform: xhs|douyin|kuaishou|bilibili|weibo|tieba|zhihu")
    type: str = Field(description="Crawler type: search|detail|creator|comment")
    config: dict = Field(description="Task configuration")


class TaskCreate(TaskBase):
    """Task creation schema"""
    pass


class TaskResponse(TaskBase):
    """Task response schema"""
    id: str
    status: str
    start_time: Optional[datetime] = Field(alias="startTime", default=None)
    end_time: Optional[datetime] = Field(alias="endTime", default=None)
    items_collected: int = Field(alias="itemsCollected", default=0)
    progress: int = Field(default=0)
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")
    
    class Config:
        populate_by_name = True
        from_attributes = True


class TaskDetail(TaskResponse):
    """Detailed task response with logs and errors"""
    duration: Optional[int] = Field(default=None, description="Duration in seconds")
    logs: list[TaskLog] = Field(default_factory=list)
    errors: list[dict] = Field(default_factory=list)
    
    class Config:
        populate_by_name = True
        from_attributes = True


class TaskListItem(BaseModel):
    """Task list item schema"""
    id: str
    name: Optional[str] = None
    platform: str
    type: str = Field(alias="crawlerType")
    status: str
    start_time: Optional[datetime] = Field(alias="startTime", default=None)
    estimated_time: Optional[int] = Field(alias="estimatedTime", default=None)
    progress: int
    items_collected: int = Field(alias="itemsCollected")
    config: dict
    created_at: datetime = Field(alias="createdAt")
    
    class Config:
        populate_by_name = True
        from_attributes = True
